### PR TITLE
[SPARK-11959] [SPARK-15484] [Doc] [ML] Document WLS and IRLS

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?
- Document `WeightedLeastSquares`(normal equation) and `IterativelyReweightedLeastSquares`.
- Copy `L-BFGS` documents from `spark.mllib` to `spark.ml`.

Due to the session `Optimization of linear methods` is used for developers, I think we should provide the brief introduction of the optimization method, necessary references and how it implements in Spark. It's not necessary to paste all mathematical formula and derivation here. If developers/users want to learn more, they can track reference. 
## How was this patch tested?

Document update, no tests.

### CODE REVIEW

File: docs/ml-advanced.md [hunk 1]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+* $w_i$ the weight of i-th observation
+* $a_i$ the features vector of i-th observation
+* $b_i$ the label of i-th observation
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{i=1}^n w_i} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
```
Comment (yanboliang) [inline docs/ml-advanced.md]: image

File: docs/ml-advanced.md [hunk 2]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Given $n$ weighted observations $(w_i, a_i, b_i)$:
+
+* $w_i$ the weight of i-th observation
+* $a_i$ the features vector of i-th observation
+* $b_i$ the label of i-th observation
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{i=1}^n w_i} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
```
Comment (yanboliang) [inline docs/ml-advanced.md]: image

File: docs/ml-advanced.md [hunk 3]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+# Optimization of linear methods (developer)
+
+## Limited-memory BFGS (L-BFGS)
+[L-BFGS](http://en.wikipedia.org/wiki/Limited-memory_BFGS) is an optimization 
+algorithm in the family of quasi-Newton methods to solve the optimization problems of the form 
+`$\min_{\wv \in\R^d} \; f(\wv)$`. The L-BFGS method approximates the objective function locally as a 
+quadratic without evaluating the second partial derivatives of the objective function to construct the 
+Hessian matrix. The Hessian matrix is approximated by previous gradient evaluations, so there is no 
+vertical scalability issue (the number of training features) when computing the Hessian matrix 
+explicitly in Newton's method. As a result, L-BFGS often achieves rapider convergence compared with
```
Comment (BenFradet) [inline docs/ml-advanced.md]: Rapider -> faster

File: docs/ml-advanced.md [hunk 4]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+## Limited-memory BFGS (L-BFGS)
+[L-BFGS](http://en.wikipedia.org/wiki/Limited-memory_BFGS) is an optimization 
+algorithm in the family of quasi-Newton methods to solve the optimization problems of the form 
+`$\min_{\wv \in\R^d} \; f(\wv)$`. The L-BFGS method approximates the objective function locally as a 
+quadratic without evaluating the second partial derivatives of the objective function to construct the 
+Hessian matrix. The Hessian matrix is approximated by previous gradient evaluations, so there is no 
+vertical scalability issue (the number of training features) when computing the Hessian matrix 
+explicitly in Newton's method. As a result, L-BFGS often achieves rapider convergence compared with 
+other first-order optimization.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: Optimizations

File: docs/ml-advanced.md [hunk 5]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
 
-The optimization algorithm underlying the implementation is called
 [Orthant-Wise Limited-memory
 QuasiNewton](http://research-srv.microsoft.com/en-us/um/people/jfgao/paper/icml07scalable.pdf)
-(OWL-QN). It is an extension of L-BFGS that can effectively handle L1
-regularization and elastic net.
+(OWL-QN) is an extension of L-BFGS that can effectively handle L1 regularization and elastic net.
+
+L-BFGS was used as solver for [LinearRegression](api/scala/index.html#org.apache.spark.ml.regression.LinearRegression),
```
Comment (BenFradet) [inline docs/ml-advanced.md]: Is used as a solver

File: docs/ml-advanced.md [hunk 6]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+[LogisticRegression](api/scala/index.html#org.apache.spark.ml.classification.LogisticRegression),
+[AFTSurvivalRegression](api/scala/index.html#org.apache.spark.ml.regression.AFTSurvivalRegression)
+and [MultilayerPerceptronClassifier](api/scala/index.html#org.apache.spark.ml.classification.MultilayerPerceptronClassifier).
+
+`spark.ml` L-BFGS solver calls the corresponding implementation in [breeze](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/optimize/LBFGS.scala).
```
Comment (BenFradet) [inline docs/ml-advanced.md]: The spark.ml ...

File: docs/ml-advanced.md [hunk 7]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{i=1}^n w_i} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass to collect necessary statistics to solve this function.
+Unlike the original dataset which can only be stored in distributed system,
```
Comment (BenFradet) [inline docs/ml-advanced.md]: Distributed storage

File: docs/ml-advanced.md [hunk 8]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+* $b_i$ the label of i-th observation
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{i=1}^n w_i} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass to collect necessary statistics to solve this function.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: One pass over the data to collect the necessary statistics to solve it

File: docs/ml-advanced.md [hunk 9]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory of a single machine, and then we can solve the objective function through Cholesky factorization on driver.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: "On a single machine" or "the memory of a single machine"

Comment (BenFradet) [inline docs/ml-advanced.md]: On the driver

File: docs/ml-advanced.md [hunk 10]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares only supports the number of features is no more than 4096.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: Only supports 4096 or less features

File: docs/ml-advanced.md [hunk 11]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+## Iteratively re-weighted least squares (IRLS)
+
+`spark.ml` implements the method of iteratively reweighted least squares (IRLS) by [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala).
```
Comment (BenFradet) [inline docs/ml-advanced.md]: I'd remove "the method"

File: docs/ml-advanced.md [hunk 12]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+It can be used to find maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
+Refer to [Iteratively Reweighted Least Squares for Maximum Likelihood Estimation, and some Robust and Resistant Alternatives](http://www.jstor.org/stable/2345503) for more information.
+
+It solves certain optimization problems by an iterative way:
```
Comment (BenFradet) [inline docs/ml-advanced.md]: By an iterative way -> iteratively

File: docs/ml-advanced.md [hunk 13]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory of a single machine, and then we can solve the objective function through Cholesky factorization on driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares only supports the number of features is no more than 4096.
+
+## Iteratively re-weighted least squares (IRLS)
+
+`spark.ml` implements the method of iteratively reweighted least squares (IRLS) by [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala).
+It can be used to find maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: The maximum likelihood...

**Comment by BenFradet** [comment]
@yanboliang lgtm except for a few phrasing issues.

File: docs/ml-advanced.md [hunk 14]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Refer to [Iteratively Reweighted Least Squares for Maximum Likelihood Estimation, and some Robust and Resistant Alternatives](http://www.jstor.org/stable/2345503) for more information.
+
+It solves certain optimization problems by an iterative way:
+
+* linearize the objective at current solution and update corresponding weight.
```
Comment (BenFradet) [inline docs/ml-advanced.md]: I'm not too sure about this sentence, would it be: Add  linearize the objective function at the current solution and update the corresponding weights?

**Comment by yanboliang** [comment]
@BenFradet Thanks for your comments, updated.

**Comment by jkbradley** [comment]
Taking a final look

File: docs/ml-advanced.md [hunk 15]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+# Optimization of linear methods (developer)
+
+## Limited-memory BFGS (L-BFGS)
+[L-BFGS](http://en.wikipedia.org/wiki/Limited-memory_BFGS) is an optimization 
+algorithm in the family of quasi-Newton methods to solve the optimization problems of the form 
+`$\min_{\wv \in\R^d} \; f(\wv)$`. The L-BFGS method approximates the objective function locally as a 
+quadratic without evaluating the second partial derivatives of the objective function to construct the 
+Hessian matrix. The Hessian matrix is approximated by previous gradient evaluations, so there is no 
+vertical scalability issue (the number of training features) when computing the Hessian matrix
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "when computing" ---> "unlike computing"

File: docs/ml-advanced.md [hunk 16]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
-(OWL-QN). It is an extension of L-BFGS that can effectively handle L1
-regularization and elastic net.
+(OWL-QN) is an extension of L-BFGS that can effectively handle L1 regularization and elastic net.
+
+L-BFGS is used as a solver for [LinearRegression](api/scala/index.html#org.apache.spark.ml.regression.LinearRegression),
+[LogisticRegression](api/scala/index.html#org.apache.spark.ml.classification.LogisticRegression),
+[AFTSurvivalRegression](api/scala/index.html#org.apache.spark.ml.regression.AFTSurvivalRegression)
+and [MultilayerPerceptronClassifier](api/scala/index.html#org.apache.spark.ml.classification.MultilayerPerceptronClassifier).
+
+The `spark.ml` L-BFGS solver calls the corresponding implementation in [breeze](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/optimize/LBFGS.scala).
```
Comment (jkbradley) [inline docs/ml-advanced.md]: Link to their docs, rather than code?

File: docs/ml-advanced.md [hunk 17]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+## Normal equation solver for weighted least squares (normal)
+
+`spark.ml` implements normal equation solver for weighted least squares by [WeightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala).
```
Comment (jkbradley) [inline docs/ml-advanced.md]: I'd prefer a link to Wikipedia than the source code, or both if you want to keep the source link.

File: docs/ml-advanced.md [hunk 18]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares only supports the number of features is no more than 4096.
```
Comment (jkbradley) [inline docs/ml-advanced.md]: In order to limit communication, WeightedLeastSquares requires that the number of features be no more than 4096.  For larger problems, use L-BFGS instead.

File: docs/ml-advanced.md [hunk 19]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+## Iteratively re-weighted least squares (IRLS)
+
+`spark.ml` implements iteratively reweighted least squares (IRLS) by [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala).
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "spark.ml" --> "The DataFrames-based API" or just "MLlib"
Also, how about a Wikipedia link instead of source code?

File: docs/ml-advanced.md [hunk 20]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+It can be used to find the maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
+Refer to [Iteratively Reweighted Least Squares for Maximum Likelihood Estimation, and some Robust and Resistant Alternatives](http://www.jstor.org/stable/2345503) for more information.
+
+It solves certain optimization problems iteratively:
+
+* linearize the objective at current solution and update corresponding weight.
+* solve a weighted least squares (WLS) problem by WeightedLeastSquares.
+* repeat above steps until convergence.
+
+Due to it involves solving a weighted least squares (WLS) problem by WeightedLeastSquares in each step of the iteration,
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "Due to it" --> "Since it"
"each step of the iteration" --> "each iteration"

File: docs/ml-advanced.md [hunk 21]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+it also only supports the number of features is no more than 4096.
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "also only supports the number of features is no more than 4096" --> "also requires the number of features to be no more than 4096"

File: docs/ml-advanced.md [hunk 22]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Currently IRLS was used as the default solver of [GeneralizedLinearRegression](api/scala/index.html#org.apache.spark.ml.regression.GeneralizedLinearRegression).
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "was" --> "is"

**Comment by jkbradley** [comment]
Done with pass

File: docs/ml-advanced.md (hunk 16 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: There is no docs for breeze LBFGS, the Wikipedia page for L-BFGS is linked above. I think it makes sense to link code here because most audience of the optimization session is developers or expert users.

File: docs/ml-advanced.md (hunk 18 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: I think limit communication is only one of the reasons, the most important one is that we can not solve Cholesky factorization in a single machine if the number of features more than 4096.

**Comment by yanboliang** [comment]
@jkbradley Thanks for your comments. I think most of the audiences to the optimization session are developers and expert users, so I linked both Wikipedia docs and source code to the corresponding concept. Any more comments, please feel free to let me know.

File: docs/ml-advanced.md [hunk 23]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+It can be used to find the maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
+Refer to [Iteratively Reweighted Least Squares for Maximum Likelihood Estimation, and some Robust and Resistant Alternatives](http://www.jstor.org/stable/2345503) for more information.
+
+It solves certain optimization problems iteratively:
+
+* linearize the objective at current solution and update corresponding weight.
+* solve a weighted least squares (WLS) problem by WeightedLeastSquares.
+* repeat above steps until convergence.
+
+Since it involves solving a weighted least squares (WLS) problem by WeightedLeastSquares in each iteration,
```
Comment (sethah) [inline docs/ml-advanced.md]: While this is true, it does not provide any sort of explanation as to _why_ that restriction exists. I like the idea of explaining that the covariance matrix can fit into main memory with < 4096 features (usually).

File: docs/ml-advanced.md (hunk 16 above)
Comment (jkbradley) [inline docs/ml-advanced.md]: OK, sounds fine

File: docs/ml-advanced.md (hunk 18 above)
Comment (jkbradley) [inline docs/ml-advanced.md]: True, it's also for the decomposition.  I'd rephrase at least: "In order to take the normal equation approach efficiently" --> "In order to make the normal equation approach efficient"

**Comment by jkbradley** [comment]
@yanboliang Thanks for the updates.  I responded above about a tiny fix.

File: docs/ml-advanced.md [hunk 24]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+L-BFGS is used as a solver for [LinearRegression](api/scala/index.html#org.apache.spark.ml.regression.LinearRegression),
+[LogisticRegression](api/scala/index.html#org.apache.spark.ml.classification.LogisticRegression),
+[AFTSurvivalRegression](api/scala/index.html#org.apache.spark.ml.regression.AFTSurvivalRegression)
+and [MultilayerPerceptronClassifier](api/scala/index.html#org.apache.spark.ml.classification.MultilayerPerceptronClassifier).
+
+MLlib L-BFGS solver calls the corresponding implementation in [breeze](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/optimize/LBFGS.scala).
+
+## Normal equation solver for weighted least squares (normal)
+
+MLlib implements normal equation solver for [weighted least squares](https://en.wikipedia.org/wiki/Least_squares#Weighted_least_squares) by [WeightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala).
```
Comment (sethah) [inline docs/ml-advanced.md]: We should replace "MLlib" with `spark.ml` or similar here and elsewhere.

File: docs/ml-advanced.md [hunk 25]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares requires that the number of features be no more than 4096. For larger problems, use L-BFGS instead.
+
+## Iteratively re-weighted least squares (IRLS)
+
+MLlib implements [iteratively reweighted least squares (IRLS)](https://en.wikipedia.org/wiki/Iteratively_reweighted_least_squares) by [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala).
+It can be used to find the maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
```
Comment (sethah) [inline docs/ml-advanced.md]: What is the point of explaining what IRLS _could_ be used for, even though it is not used for that in Spark? If it were a public interface, I could understand giving examples of other ways it could be used, but as things are currently I'm not sure this is necessary. If anything, it might seem like we're suggesting users can utilize IRLS for these other problems, but we don't provide a public API for it, which is misleading.

File: docs/ml-advanced.md [hunk 26]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+vertical scalability issue (the number of training features) unlike computing the Hessian matrix 
+explicitly in Newton's method. As a result, L-BFGS often achieves faster convergence compared with 
+other first-order optimizations.
 
-The optimization algorithm underlying the implementation is called
 [Orthant-Wise Limited-memory
 QuasiNewton](http://research-srv.microsoft.com/en-us/um/people/jfgao/paper/icml07scalable.pdf)
-(OWL-QN). It is an extension of L-BFGS that can effectively handle L1
-regularization and elastic net.
+(OWL-QN) is an extension of L-BFGS that can effectively handle L1 regularization and elastic net.
```
Comment (sethah) [inline docs/ml-advanced.md]: nit: "L1 and elastic net regularization"

Comment (sethah) [inline docs/ml-advanced.md]: Also, this sentence seems out of place. We don't mention that it is also used in some Spark algorithms or where it is used. Perhaps we can give it its own section, or otherwise just make a small mention that it is a variant of LBFGS that spark uses for L1 regularization algorithms.

File: docs/ml-advanced.md [hunk 27]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+`$\min_{\wv \in\R^d} \; f(\wv)$`. The L-BFGS method approximates the objective function locally as a 
+quadratic without evaluating the second partial derivatives of the objective function to construct the 
+Hessian matrix. The Hessian matrix is approximated by previous gradient evaluations, so there is no 
+vertical scalability issue (the number of training features) unlike computing the Hessian matrix 
+explicitly in Newton's method. As a result, L-BFGS often achieves faster convergence compared with 
+other first-order optimizations.
 
-The optimization algorithm underlying the implementation is called
 [Orthant-Wise Limited-memory
 QuasiNewton](http://research-srv.microsoft.com/en-us/um/people/jfgao/paper/icml07scalable.pdf)
```
Comment (sethah) [inline docs/ml-advanced.md]: minor: "Quasi-Newton"

File: docs/ml-advanced.md [hunk 28]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares requires that the number of features be no more than 4096. For larger problems, use L-BFGS instead.
+
+## Iteratively re-weighted least squares (IRLS)
```
Comment (sethah) [inline docs/ml-advanced.md]: I don't really have a preference for "re-weighted" vs "reweighted", but we should be consistent.

File: docs/ml-advanced.md [hunk 29]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+MLlib implements [iteratively reweighted least squares (IRLS)](https://en.wikipedia.org/wiki/Iteratively_reweighted_least_squares) by [IterativelyReweightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/IterativelyReweightedLeastSquares.scala).
+It can be used to find the maximum likelihood estimates of a generalized linear model (GLM), find M-estimator in robust regression and other optimization problems.
+Refer to [Iteratively Reweighted Least Squares for Maximum Likelihood Estimation, and some Robust and Resistant Alternatives](http://www.jstor.org/stable/2345503) for more information.
+
+It solves certain optimization problems iteratively:
```
Comment (sethah) [inline docs/ml-advanced.md]: "It solves certain optimization problems iteratively through the following procedure:" ? It makes it more clear what the bullet points below are explaining.

File: docs/ml-advanced.md (hunk 24 above)
Comment (sethah) [inline docs/ml-advanced.md]: Ah, well I see this was suggested above. I think it's a bit confusing to say "MLlib", but I will defer to others.

File: docs/ml-advanced.md (hunk 23 above)
Comment (sethah) [inline docs/ml-advanced.md]: Maybe I was looking at an outdated diff before, I see now that you've explained this above. Maybe add a "see above for further information on this constraint" ?

File: docs/ml-advanced.md [hunk 30]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+(OWL-QN) is an extension of L-BFGS that can effectively handle L1 regularization and elastic net.
+
+L-BFGS is used as a solver for [LinearRegression](api/scala/index.html#org.apache.spark.ml.regression.LinearRegression),
+[LogisticRegression](api/scala/index.html#org.apache.spark.ml.classification.LogisticRegression),
+[AFTSurvivalRegression](api/scala/index.html#org.apache.spark.ml.regression.AFTSurvivalRegression)
+and [MultilayerPerceptronClassifier](api/scala/index.html#org.apache.spark.ml.classification.MultilayerPerceptronClassifier).
+
+MLlib L-BFGS solver calls the corresponding implementation in [breeze](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/optimize/LBFGS.scala).
+
+## Normal equation solver for weighted least squares (normal)
```
Comment (sethah) [inline docs/ml-advanced.md]: What is the "(normal)" supposed to designate?

File: docs/ml-advanced.md [hunk 31]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
+Unlike the original dataset which can only be stored in distributed system,
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
+In order to take the normal equation approach efficiently, WeightedLeastSquares requires that the number of features be no more than 4096. For larger problems, use L-BFGS instead.
```
Comment (sethah) [inline docs/ml-advanced.md]: Does it make sense to direct readers to "use L-BFGS" instead? Spark doesn't expose an L-BFGS api nor does it even expose the normal equation solver to users.

File: docs/ml-advanced.md [hunk 32]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{i=1}^n w_i} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
+Unlike the original dataset which can only be stored in distributed system,
```
Comment (sethah) [inline docs/ml-advanced.md]: minor: "in distributed fashion" or "in a distributed system".

File: docs/ml-advanced.md [hunk 33]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
+
+WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization, standardizing features and labels.
```
Comment (sethah) [inline docs/ml-advanced.md]: This sentence is quite confusing to me. What is the "standardizing features and labels" part supposed to mean? Should it read: "... and provides options to enable or disable regularization and standardization" ?

File: docs/ml-advanced.md (hunk 18 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: Sounds good, update.

File: docs/ml-advanced.md [hunk 34]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+Given $n$ weighted observations $(w_i, a_i, b_i)$:
+
+* $w_i$ the weight of i-th observation
+* $a_i$ the features vector of i-th observation
+* $b_i$ the label of i-th observation
+
+The number of features for each observation is $m$. We use the following weighted least squares formulation:
+`\[   
+minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{k=1}^n w_k} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\]`
```
Comment (yanboliang) [inline docs/ml-advanced.md]: image

File: docs/ml-advanced.md (hunk 23 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: Since we mentioned IRLS called WLS, it's straight forward for users to refer WLS for the reason of the constraint.

File: docs/ml-advanced.md (hunk 24 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: We will switch RDD-based MLlib APIs to maintenance mode in Spark 2.0. MLlib mainly refer to the DataFrame-based API later, so I think it's OK to use MLlib.

File: docs/ml-advanced.md (hunk 25 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: I think most audience of this session is developers rather than users as you see in the title ```Optimization of linear methods (developer)```, so I explain the design considerations behind the IRLS implementation. Developers who have ideas to implement algorithms based on IRLS can contribute to Spark MLlib.

File: docs/ml-advanced.md (hunk 30 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: We have different ```solver``` for linear models for users to set, and the words in brackets is the name of the solver.

File: docs/ml-advanced.md (hunk 31 above)
Comment (yanboliang) [inline docs/ml-advanced.md]: Yes, we do not expose L-BFGS api, but we exposed the param ```solver``` which can control the optimization method used to solve the linear model.

**Comment by yanboliang** [comment]
@sethah @jkbradley Thanks for your comments! I addressed most of the comments and responded inline.

File: docs/ml-advanced.md (hunk 25 above)
Comment (sethah) [inline docs/ml-advanced.md]: I had understood "developer" to designate developers of Spark applications, not developers who contribute to the core Apache Spark project. For example, we designate some public APIs as "developer API". At any rate, I'm fine leaving this.

File: docs/ml-advanced.md (hunk 30 above)
Comment (sethah) [inline docs/ml-advanced.md]: So this is supposed to be the actual `String` value they should set the solver param to if they want to use it? I don't see the point of including that here since we document in the API param docs, and it could change and leave the user guide out of sync. Further, I don't think most people will realize that is what those mean. It might be better to put them in quotes (e.g. "normal"), but I'd vote for not having them at all.

File: docs/ml-advanced.md [hunk 35]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of label
```
Comment (sethah) [inline docs/ml-advanced.md]: minor: "the label"

File: docs/ml-advanced.md [hunk 36]
```diff
@@ -4,10 +4,85 @@ title: Advanced topics - spark.ml
...
+and $\sigma_j$ is the population standard deviation of the j-th feature column.
+
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
+Unlike the original dataset which can only be stored in a distributed system,
+these statistics can be easily loaded into memory on a single machine, and then we can solve the objective function through Cholesky factorization on the driver.
```
Comment (sethah) [inline docs/ml-advanced.md]: This isn't true in general. We should say that these statistics can be loaded into memory _if the number of features is relatively small_. It might be good to indicate that the size of these statistics depends on `numFeatures^2`.

File: docs/ml-advanced.md (hunk 34 above)
Comment (sethah) [inline docs/ml-advanced.md]: Using `\min_x` is probably more appropriate notation.

**Comment by sethah** [comment]
Should we point the section in MLlib to the section here now? Otherwise we leave some text that is duplicated verbatim in the user guides. 

One other minor comment and this LGTM.

**Comment by jkbradley** [comment]
I think it's OK to leave the duplicate text in the MLlib guide.  It will just sit there unchanging and should not require extra effort to maintain.

**Comment by jkbradley** [comment]
LGTM
Merging with master and branch-2.0
Thanks @yanboliang @BenFradet @sethah !
